### PR TITLE
format ./python with `black`

### PR DESCRIPTION
**Pull request by manzt**
Looping both of you in case either oppose.

This PR adds `black` to our work flow as detailed in #85.

### CODE REVIEW

**Comment by ilan-gold** [comment]
Yes, I am in favor of this.

File: python/cell_reader.py [hunk 1]
```diff
@@ -315,28 +329,19 @@ def add_pca(metadata):
...
         print(spaced_clusters_json, file=args.clusters_file)
 
     if args.genes_file:
         genes = get_genes(metadata)
         genes_json = json.dumps(genes)
-        spaced_genes_json = genes_json.replace(
-            '},',
-            '},\n'
-        )
+        spaced_genes_json = genes_json.replace("},", "},\n")
```
Comment (mccalluc) [inline python/cell_reader.py:338]: Something that was easy to read at a glance is harder to read.

File: python/cytokit_reader.py [hunk 2]
```diff
@@ -10,39 +10,39 @@ def round_conv(s):
...
     parser.add_argument(
-        '--cytokit', required=True,
-        help='CSV from Cytokit')
-    parser.add_argument(
-        '--cells_file', type=argparse.FileType('x'),
-        help='Write the cleaned cell data to this file.')
+        "--cells_file",
+        type=argparse.FileType("x"),
+        help="Write the cleaned cell data to this file.",
+    )
```
Comment (mccalluc) [inline python/cytokit_reader.py:45]: Parallelism in argument structures is no longer as obvious.

File: python/delaunay.py [hunk 3]
```diff
@@ -135,13 +137,19 @@ def distanceSquare(pt):
             self.boundaryEdges.add(e20)
 
             e01 = self.makeKey(e01[0], e01[1])
-            self.edge2Triangles[e01] = [0, ]
+            self.edge2Triangles[e01] = [
+                0,
+            ]
```
Comment (mccalluc) [inline python/delaunay.py:142]: Huh, I'm not sure about the original intention here: perhaps it was supposed to be a tuple? In any case, I don't think this makes it more clear.

File: python/dries_json_reader.py [hunk 4]
```diff
@@ -69,55 +69,62 @@ def factors_json(data):
...
-        factors = data[cell_id]['factors']
+        factors = data[cell_id]["factors"]
         pleiden_cells[cell_id] = pleiden_clusters.index(
-            factors['pleiden_clus'][0]
-        )
-        kmeans_cells[cell_id] = kmeans_clusters.index(
-            factors['kmeans'][0]
+            factors["pleiden_clus"][0]
         )
+        kmeans_cells[cell_id] = kmeans_clusters.index(factors["kmeans"][0])
```
Comment (mccalluc) [inline python/dries_json_reader.py:92]: Parallelism less clear

File: python/loom_reader.py [hunk 5]
```diff
@@ -6,39 +6,45 @@ class LoomReader:
...
+        for (
+            valid,
+            cell_id,
+            tsne1,
+            tsne2,
+            cluster_id,
+            cluster_name,
+            x,
+            y,
+        ) in data_zip:
```
Comment (mccalluc) [inline python/loom_reader.py:36]: Sometimes there is a logical grouping within a larger list: Flattening it all the way out isn't more clear.

**Review by mccalluc — COMMENTED** [review]
This wouldn't be my preference. It *definitely* needs linting, (sorry not to have done that earlier!) and there are sections that are badly formatted, but I think applying black looses information that can make the code more readable.

I've flagged some cases in the diff: Generally:
- Sometimes parallel structures are obscured because black can fit all of one case on one line, but another needs to be split.
- Sometimes logical groupings in a list are obscured because black can either fit it on one line, or split it onto multiple.
- Where comments go when lines are rearranged might not be intuitive.
- Sometimes, particularly with string replacement, even short single arguments should be on their own lines for clarity.

(I also prefer single quotes, but that's my least persuasive argument, and black can allow it, in any event.)

If we were wasting time on formatting debates, perhaps, but I think the judicious use of whitespace can make code more clear, and I'm reluctant to give that up.

So: not going to say "over my dead body", but could you mull this over for a couple days? Ultimately, you guys are more active than me in this repo, and if black makes your work easier, I won't block, but I'm not really comfortable with the trade-offs this incurs.

**Comment by manzt** [comment]
Thanks for having a look over this. Sorry for making you go through the diff, and I think you bring up some valid points. 

I think adding some formatting to our workflow would unify the code-base, but I agree now that running this blindly is probably not the best approach. Thanks for all the examples (and again sorry for making you do that).

### preserving logical structure / parallelism
Preserving logical structure and parallelism is the most compelling to me. These seem to be most abundant with the `argsparse` options in our code base. I propose that we use`# fmt: off` and `# fmt: on` blocks to preserve these multi-line sections, and make the "judicious use of whitespace" explicit:

```python
        # fmt: off
        for (valid, cell_id, tsne1, tsne2,
             cluster_id, cluster_name, x, y) in data_zip:
        # fmt: on
```

I don't think that this will clutter the codebase much since the blocks are often multi-line and not highly abundant. Plus the `# fmt: off/on` blocks will be indicative to the reader that the use of whitespace _is_ important, or that there is logical structure within the block. From experience, when adding additional arguments with `parser.add_argument` in the past, I did not know to adhere to this structure. I'd suggest adding `--diff` flag to `./test.sh`:

```
# ./test.sh
start black
black --line-length 79 -diff --check ./python
end black
```

So that any changes black would make are communicated to the user.

### line wrapping
I think differences with some single-line cases mentioned above come from black trying to make a decision of what to do with a strict line-length of 79. From the docs "how black wraps lines":

> "The rules for horizontal whitespace can be summarized as: do whatever makes `pycodestyle` happy. The coding style used by Black can be viewed as a strict subset of PEP 8."

In my experience using black, the default line length of 88 is the most readable. 

> "You probably noticed the peculiar default line length. Black defaults to 88 characters per line, which happens to be 10% over 80. This number was found to produce significantly shorter files than sticking with 80 (the most popular), or even 79 (used by the standard library). In general, [90-ish seems like the wise choice](https://www.youtube.com/watch?v=wf-BqAjZb8M&feature=youtu.be&t=260)."

However, to get this to work with the linter we would need to add:

```
# ./test.sh
start flake8
flake8 --max-line-length 88 --exclude ./vips-8.7.1/
end flake8
```

which I don't mind.

### final thoughts

I'm also not putting my foot down here, but am trying to see if we can come to some compromise. I agree that preserving thoughtful use of whitespace is import and I don't want to remove that, but I'd rather be explicit in these situations and otherwise adhere to some opinionated "standard" that I don't need to think about. My preferences for defaults would be:

```
# ./test.sh
start black
black --diff --check ./python
end black

start flake8
flake8 --max-line-length 88 --exclude ./vips-8.7.1/
end flake8
```

Black does allow for single quotes with the `-S/--skip-string-normalization` flag. However, it doesn't allow you to enforce a quoting style other than double. I too also prefer single quotes generally, but I like the strictness of an opinionated formatter more. I'm happy to use this flag, but I'd like to agree "we are using single quotes", and then it's fair to make the suggestion in code reviews.

**Comment by manzt** [comment]
Those are my thoughts as of this morning. I have other things to work on and will put this in the back of my mind for a couple days. Thanks again for your comments.

**Comment by mccalluc** [comment]
Cool: Thank you for taking the time to write up those thoughts. I'm not sure whether we'd really remember to use `# fmt: off` when it would be useful. (When I used black, briefly, I didn't realize what it was doing until I looked at the diff.)

So let's put this on hold till Friday, talk it over, and if you guys what to try it, I won't block.

**Pull request closed without merging**